### PR TITLE
Test location-based user edit restrictions

### PR DESCRIPTION
**Pull request by esoergel**
Follow-up from https://github.com/dimagi/commcare-hq/pull/13095
@czue @millerdev

### CODE REVIEW

File: corehq/apps/locations/tests/test_permissions.py [hunk 1]
```diff
@@ -197,3 +208,34 @@ def test_cant_edit_parent_location(self):
...
+        self._assert_edit_user_gives_status(self.boston_worker, 200)
+
+    def test_cant_edit_worker(self):
+        self._assert_edit_user_gives_status(self.cambridge_worker, 404)
+
+    def test_restricted_worker_list(self):
+        url = reverse(user_views.MobileWorkerListView.urlname, args=[self.domain])
+        self.client.login(username=self.suffolk_user.username, password="password")
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, status_code)
```
Comment (dimagimon) [inline corehq/apps/locations/tests/test_permissions.py]: F821 undefined name 'status_code'

File: corehq/apps/locations/tests/test_permissions.py [hunk 2]
```diff
@@ -234,8 +243,31 @@ def test_can_edit_worker(self):
...
+        data = json.dumps(data)
+        return self.client.post(
+            url,
+            content_type="application/json;charset=utf-8",
+            **{
+                'HTTP_DJNG_REMOTE_METHOD': method_name,
+                'HTTP_X_REQUESTED_WITH': 'XMLHttpRequest',
+                'CONTENT_LENGTH': len(data),
+                'wsgi.input': StringIO(data),
+            })
```
Comment (esoergel) [inline corehq/apps/locations/tests/test_permissions.py]: This magic combination of parameters and serialization of data was incredibly obnoxious to figure out.  If anyone else ever needs to test djangoRMI methods, hopefully they'll find this.

**Comment by millerdev** [comment]
Tests failed. One node failed with this:
```
FAIL: test_get_choices_for_values (corehq.apps.userreports.tests.test_choice_provider.UserChoiceProviderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_choice_provider.py", line 245, in test_get_choices_for_values
    ['unknown-user'] + [user._id for user in self.users])
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_choice_provider.py", line 102, in _test_get_choices_for_values
    self.static_choice_provider.get_choices_for_values(values),
AssertionError: Items in the first set but not the second:
Choice(value=None, display='')
Items in the second set but not the first:
Choice(value=None, display=u'dennis')
Choice(value=None, display=u'albert')
Choice(value=None, display=u'bernice')
Choice(value=None, display=u'elizabeth')
Choice(value=None, display=u'candice@example.com'
```

and two others failures:
```
ERROR: test suite for <class 'corehq.apps.reports.tests.test_project_health.ProjectHealthDashboardTest'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/vendor/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 179, in setUpClass
    cls.class_setup()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 33, in class_setup
    cls._setup_malt_tables()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 70, in _setup_malt_tables
    app_id=MISSING_APP_ID,
  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 372, in create
    obj.save(force_insert=True, using=self.db)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 589, in save
    force_update=force_update, update_fields=update_fields)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 617, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 698, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 731, in _do_insert
    using=using, raw=raw)
  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 921, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/vendor/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 921, in execute_sql
    cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
IntegrityError: null value in column "user_id" violates not-null constraint
DETAIL:  Failing row contains (1, 2016-10-01, null, testuser, , test_domain, 16, _MISSING_APP_ID, f, null, null, CommCareUser, 15, null, 3).
```

```
ERROR: test suite for <class 'corehq.apps.reports.tests.test_project_health.MonthlyPerformanceSummaryTests'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/vendor/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 100, in setUpClass
    cls.class_setup()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 33, in class_setup
    cls._setup_malt_tables()
  File "/mnt/commcare-hq-ro/corehq/apps/reports/tests/test_project_health.py", line 70, in _setup_malt_tables
    app_id=MISSING_APP_ID,
  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 372, in create
    obj.save(force_insert=True, using=self.db)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 589, in save
    force_update=force_update, update_fields=update_fields)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 617, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 698, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/vendor/lib/python2.7/site-packages/django/db/models/base.py", line 731, in _do_insert
    using=using, raw=raw)
  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 921, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/vendor/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 921, in execute_sql
    cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
IntegrityError: null value in column "user_id" violates not-null constraint
DETAIL:  Failing row contains (1, 2016-10-01, null, testuser, , test_domain, 16, _MISSING_APP_ID, f, null, null, CommCareUser, 15, null, 3).
```

kicked all three just in case they're transient issues.

**Comment by millerdev** [comment]
👍 on ✅

**Comment by czue** [comment]
looks good. out of curiosity why do you prefer ESFakes over just calling ES directly?

**Comment by esoergel** [comment]
No particular reason - I was having trouble getting the ES tests to run locally, and I remembered the ESFakes thing, and figured since I wasn't really testing the ES query logic, just that it applied the filter when applicable, it seemed like a good fit.

**Comment by esoergel** [comment]
:white_check_mark:

**Pull request merged**